scripts/update-nsis-functions.sh

```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
+#
+# Author:  Enrico Tröger
+# License: GPL v2 or later
+#
+# Updates the `functions` and `variables` entries in data/filetypes.nsis.
+
+set -e
+
+TOKENS_CPP_FILE="/tmp/nsis_tokens.cpp"
+TOKENS_CPP_URL="https://raw.githubusercontent.com/kichik/nsis/master/Source/tokens.cpp"
+BUILD_CPP_FILE="/tmp/nsis_build.cpp"
+BUILD_CPP_URL="https://raw.githubusercontent.com/kichik/nsis/master/Source/build.cpp"
+DATA_FILE=data/filedefs/filetypes.nsis
+
+[ -f "${DATA_FILE}" ]
+
+# download tokens.cpp and build.cpp from NSIS
+wget --quiet --output-document="${TOKENS_CPP_FILE}" "${TOKENS_CPP_URL}"
+wget --quiet --output-document="${BUILD_CPP_FILE}" "${BUILD_CPP_URL}"
+
+# extract function names (then sort the result, conver to lowercase and replace new lines by spaces)
+functions=$(\
+    sed --silent --regexp-extended 's/^\{TOK_.*,_T\("(.*)"\),[0-9]+,.*$/\1/p' "${TOKENS_CPP_FILE}" | \
+    sort | \
+    tr '[:upper:]' '[:lower:]' | \
+    tr '\n' ' ')
+
+# extract variable names (then sort the result, conver to lowercase and replace new lines by spaces)
+variables=$(\
+    sed --silent --regexp-extended\
+        --expression 's/^[ ]*m_ShellConstants.add\(_T\("(.*)"\),.*,.*\);.*$/\1/p' \
+        --expression 's/^[ ]*m_UserVarNames.add\(_T\("(.*)"\),.*\);.*$/\1/p' "${BUILD_CPP_FILE}" | \
+    sort | \
+    tr '[:upper:]' '[:lower:]' | \
+    tr '\n' ' ')
+
+# hardcode a few more, as found in the documentation ("4.2.2 Other Writable Variables")
+variables_extra='{nsisdir} 0 1 2 3 4 5 6 7 8 9 r0 r1 r2 r3 r4 r5 r6 r7 r8 r9 \\n \\r \\t $'
+variables="${variables_extra} ${variables}"
+# remove leading and trailing whitespaces
+variables="$(echo "$variables" | sed 's/^[[:blank:]]*//;s/[[:blank:]]*$//')"
+functions="$(echo "$functions" | sed 's/^[[:blank:]]*//;s/[[:blank:]]*$//')"
```

Comment:
what about moving that in the initial generation next to the other ones?  also it looks like a little bit of duplication, so what about adding a small function like this:
```shell
normalize() {
   sort | tr '[:upper:]' '[:lower:]' | tr '\n' ' ' | sed 's/^[[:blank:]]*//;s/[[:blank:]]*$//'
}

functions=$(
    sed --silent --regexp-extended 's/^\{TOK_.*,_T\("(.*)"\),[0-9]+,.*$/\1/p' "${TOKENS_CPP_FILE}" | \
    normalize
)

variables=$(
    sed --silent --regexp-extended \
        --expression 's/^[ ]*m_ShellConstants.add\(_T\("(.*)"\),.*,.*\);.*$/\1/p' \
        --expression 's/^[ ]*m_UserVarNames.add\(_T\("(.*)"\),.*\);.*$/\1/p' "${BUILD_CPP_FILE}" | \
    normalize
)

# […]
```

Reply:
Nice, thanks (https://github.com/geany/geany/pull/2181/commits/342d6ffeaac3e99a89a39d4005f9121acd08f410)!